### DISCUSSION TITLE
How to change the installed location of dylibs on macOS?

### DISCUSSION THREAD

**Discussion by luckman212** [zxing-cpp/zxing-cpp#493] [Q&A] [answered]
> _related: https://github.com/zxing-cpp/zxing-cpp/issues/112_

I was able to compile zxing-cpp 2.0.0 on macOS Ventura using the commands from the README, and it runs—as long as the 3 dylibs (1 actual file and 2 symlinks) are moved to `/usr/local/lib/`

```
git clone https://github.com/zxing-cpp/zxing-cpp.git --single-branch --depth 1
cmake -S zxing-cpp -B zxing-cpp.release -DCMAKE_BUILD_TYPE=Release
cmake --build zxing-cpp.release -j8
mv -f zxing-cpp.release/core/*.dylib /usr/local/lib/
```

But, Homebrew `brew doctor` complains about those dylibs which screws up some of my scripts. 

```
Warning: Unbrewed dylibs were found in /usr/local/lib.
If you didn't put them there on purpose they could cause problems when
building Homebrew formulae, and may need to be deleted.

Unexpected dylibs:
  /usr/local/lib/libZXing.2.0.0.dylib
```

I want to put them somewhere else, but not sure what flags to pass to `cmake` or config to edit to change the linked path. I already tried editing `core/zxing.pc` and changing the `libdir=` line but that didn't work. Anyone got an idea?

**Comment by axxel** [comment]
> If you didn't put them there on purpose...

But you did, didn't you? Sounds like a homebrew issue, nothing zxing-cpp can do about it. Maybe the most elegant solution would be to come up with a homebrew formulae. This way we could even have this software officially distributed via homebrew?

**Comment by luckman212** [comment]
> _Sounds like a homebrew issue_

Maybe, but I'm really just trying to ask in general about the dynamic linking and how to change the search paths for the dylibs. If this question belongs somewhere else I'd love to learn about it.

**Comment by axxel** [comment]
This is a question that concerns your executable, not the library itself. On linux you could simply set the LD_LIBRARY_PATH environment variable before executing your binary to let the dynamic linker know about other places where libs are to be found. There should be something similiar on macOS. Also I remember playing with `otool` on macOS once to manipulate the `rpath` (if my memory serves me right) which sort of hard-codes the library path into the executable. That's it about it in terms of pointers I can offer in that regard.

**Comment by luckman212** [comment, marked as answer]
@axxel Thanks for the hint. I really struggled to figure this one out, as I found the [CMake docs on changing `RPATH`](https://gitlab.kitware.com/cmake/community/-/wikis/doc/cmake/RPATH-handling) pretty tough to untangle. I'm sure that's because of my lack of knowledge. I eventually did get a working build script sorted out. Sharing below in case it helps anyone!

By saving the script below as `build_zxing-cpp.bash` and executing `bash ./build_zxing-cpp.bash` you should end up with a working ZXingReader and ZXingWriter in your `/usr/local/bin` directory, complete with properly-linked .dylibs that don't cause Homebrew to barf.

```shell
#!/usr/bin/env bash

git clone https://github.com/zxing-cpp/zxing-cpp.git --single-branch --depth 1
cmake -S zxing-cpp -B zxing-cpp.release -DCMAKE_BUILD_TYPE=Release
cmake --build zxing-cpp.release -j8
mkdir -p /usr/local/lib/zxing-cpp 2>/dev/null
mv -f zxing-cpp.release/core/*.dylib /usr/local/lib/zxing-cpp/

OLD_RPATH=$(otool -l zxing-cpp.release/example/ZXingReader | grep LC_RPATH -A2 | sed -En 's#^[[:space:]]*path ([^ ]+).*$#\1#p')
[[ -d $OLD_RPATH ]] || { echo 1>&2 "error, OLD_RPATH not found"; exit 1; }
NEW_RPATH=/usr/local/lib/zxing-cpp

install_name_tool -rpath "$OLD_RPATH" $NEW_RPATH zxing-cpp.release/example/ZXingReader
install_name_tool -rpath "$OLD_RPATH" $NEW_RPATH zxing-cpp.release/example/ZXingWriter

cp -v zxing-cpp.release/example/ZXingReader zxing-cpp.release/example/ZXingWriter /usr/local/bin/
```

**Reply by axxel** [reply]
Thanks for sharing the script. Since this is not really a bug in the zxing-cpp project that can be 'closed' and people should be able to find it later, too, I decided to convert this into a discussion.

**Comment by luckman212** [comment]
Here's my latest build/install script for 2.1.0. In the future, you can just adjust the `LATEST_TAG` variable at the top for new releases. Everything else is automated.

You should end up with usable `ZXingReader` and `ZXingWriter` binaries in `/usr/local/bin`

```bash
#!/usr/bin/env bash

#setup
CURLOPTS=( --silent --location --connect-timeout 10 --max-time 30 )
cd "$HOME/Downloads" || exit 1

#download & extract
LATEST_TAG='v2.1.0'
FILENAME="${LATEST_TAG}.tar.gz"
if ! curl "${CURLOPTS[@]}" -o "$HOME/Downloads/$FILENAME" "https://github.com/zxing-cpp/zxing-cpp/archive/refs/tags/${LATEST_TAG}.tar.gz"; then
  exit 1
fi
tar xzf "${LATEST_TAG}.tar.gz"
cd "zxing-cpp-${LATEST_TAG#v}" || exit 1

#compile
cmake -S . -B zxing-cpp.release -DCMAKE_BUILD_TYPE=Release
cmake --build zxing-cpp.release -j8

#relink dylib paths
OLD_RPATH=$(otool -l zxing-cpp.release/example/ZXingReader | grep LC_RPATH -A2 | sed -En 's#^[[:space:]]*path ([^ ]+).*$#\1#p')
[[ -d $OLD_RPATH ]] || { echo 1>&2 "error, OLD_RPATH not found"; exit 1; }
NEW_RPATH='/usr/local/lib/zxing-cpp'
install_name_tool -rpath "$OLD_RPATH" $NEW_RPATH zxing-cpp.release/example/ZXingReader
install_name_tool -rpath "$OLD_RPATH" $NEW_RPATH zxing-cpp.release/example/ZXingWriter

#copy
cp -v zxing-cpp.release/example/ZXingReader zxing-cpp.release/example/ZXingWriter /usr/local/bin
mkdir -p /usr/local/lib/zxing-cpp
rm /usr/local/lib/zxing-cpp/*.dylib
cp -f zxing-cpp.release/core/*.dylib /usr/local/lib/zxing-cpp

#cleanup
cd ~ || exit 1
rm -rf "$HOME/Downloads/zxing-cpp-${LATEST_TAG#v}"
rm "$HOME/Downloads/${LATEST_TAG}.tar.gz"
```
